packages/http/src/mocker/index.ts

```diff
@@ -41,20 +51,52 @@ const mock: IPrismComponents<IHttpOperation, IHttpRequest, IHttpResponse, IMockH
 
       return config;
     }),
-    chain(mockConfig => negotiateResponse(mockConfig, input, resource)),
-    chain(result => assembleResponse(result, payloadGenerator))
+    Reader.chain(mockConfig => negotiateResponse(mockConfig, input, resource)),
+    Reader.chain(result => assembleResponse(result, payloadGenerator)),
+    Reader.chain(response =>
+      withLogger(logger =>
+        pipe(
+          response,
+          Either.map(r => runCallbacks({ resource, request: input.data, response: r })(logger)),
```

Comment:
Hmm, I am almost sure that running the callback is an independent operation from `negotiateResponse` and `assembleResponse ` — meaning we can throw the callbacks as soon as we have the response in place.

I do not know how much hard would be to do this (I am assuming we'd need to use a `sequenceT/S` or something; while this is not urgent, we might start thinking about this.

---

Also, given this will return a `ReaderTaskEither` — map won't be enough here anymore.

Reply:
> Hmm, I am almost sure that running the callback is an independent operation from `negotiateResponse` and `assembleResponse ` — meaning we can throw the callbacks as soon as we have the response in place.

Callbacks require the full response object of enclosing operation. See `runtimeExpression` syntax. They can access response object to retrieve certain data.

The real-life example for that is:
1) `POST /invoices` - create invoice, at request stage invoice id does not yet exists
2) response returns the invoice data together with the id
3)  payment notification callback: `{$request.body#/url}?invoiceId={$response.body#/id}` - client wants invoice id in query param because it is in java and it would require 3 classes and 6 interfaces to decode body